File: 53_MaximumSubarray.java
class Solution {
    public int maxSubArray(int[] nums) {
    	int len = nums.length;
    	if(len == 0 || nums == null) return 0;
    	
    	int sum = 0;
    	int res = Integer.MIN_VALUE;
    	
    	for(int i = 0; i < len; i++) {
    		sum = Math.max(nums[i], nums[i] + sum);
    		res = Math.max(res, sum);
    	}
    	return res;
    }
}

File: c++/53_MaximumSubarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> f(n);
        f[0] = nums[0];
        int res = f[0];
        for (int i = 1; i < n; i++) {
            f[i] = max(f[i - 1] + nums[i], nums[i]);
            res = max(res, f[i]);
        }
        return res;
    }
};
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> f(2);
        f[0 & 1] = nums[0];
        int res = f[0 & 1];
        for (int i = 1; i < n; i++) {
            f[i & 1] = max(f[i - 1 & 1] + nums[i], nums[i]);
            res = max(res, f[i & 1]);
        }
        return res;
    }
};